### ISSUE TITLE
[REGRESSION]: Performance downgrade (almost 3 times slower)

### ISSUE THREAD

**Issue by MattEarth** [microsoft/playwright#21485]
**Context:**
- GOOD Playwright Version: 1.28
- BAD Playwright Version: 1.31
- Operating System: Ubuntu 20.04, Windows
- Extra: jest runner

**Describe the bug**

Not sure, if this is proper type of "issue". However, I experienced downgrade of performance in the test execution. Between the version 1.28 and 1.31, speed of execution has been decreased almost 3 times. Not only for the workers, but also for a single run. 

Here is a piece of Jenkins log (same agent in both cases):
1.28
>[Pipeline] sh
>+ npm list playwright
>Ra.Playwright@1.0.0 /opt/jenkins/workspace/Tests/Ra.Playwright
>└── playwright@1.28.1

>[Pipeline] sh
>+ npm run support -- CI

> Ra.Playwright@1.0.0 support
> jest --verbose --maxWorkers=10 --updateSnapshot --config ./jest.config.js ./tests/support/.*.spec.js CI

>PASS browser: chromium tests/support/support.spec.js (17.505 s)
>  Support (menu item) tests
>    ✓ 01: check if links/icons/titles are displayed (1173 ms)
>    ✓ 02: check change log details (4247 ms)

> › 2 snapshots written.
>📦 report is created on: /opt/jenkins/workspace/Tests/Ra.Playwright/reports/jest_html_reporters.html
>Snapshot Summary
> › 2 snapshots written from 1 test suite.

>Test Suites: 1 passed, 1 total
>Tests:       2 passed, 2 total
>Snapshots:   2 written, 2 total
>Time:        18.434 s
>Ran all test suites matching /.\/tests\/support\/.*.spec.js|CI/i.
>[Pipeline] }
>[Pipeline] // timeout
>[Pipeline] echo
>All done!`

1.31
>[Pipeline] sh
>+ npm list playwright
>Ra.Playwright@1.0.0 /opt/jenkins/workspace/Tests/Ra.Playwright
>└── playwright@1.31.2

>[Pipeline] sh
>+ npm run support -- CI

> Ra.Playwright@1.0.0 support
> jest --verbose --maxWorkers=10 --updateSnapshot --config ./jest.config.js ./tests/support/.*.spec.js CI

>PASS browser: chromium tests/support/support.spec.js (75.641 s)
>  Support (menu item) tests
>    ✓ 01: check if links/icons/titles are displayed (33185 ms)
>    ✓ 02: check change log details (27162 ms)

> › 2 snapshots written.
>📦 report is created on: /opt/jenkins/workspace/Tests/Ra.Playwright/reports/jest_html_reporters.html
>Snapshot Summary
> › 2 snapshots written from 1 test suite.

>Test Suites: 1 passed, 1 total
>Tests:       2 passed, 2 total
>Snapshots:   2 written, 2 total
>Time:        76.184 s
>Ran all test suites matching /.\/tests\/support\/.*.spec.js|CI/i.
>[Pipeline] }
>[Pipeline] // timeout
>[Pipeline] echo
>All done!

**Comment by mxschmitt** [comment]
Can you provide us a repro which we can run locally? And more information about the environment where your tests are running? Linux distribution, version etc. Happy to look into it, but we first need a repro with test cases which surfaces it.

**Comment by mxschmitt** [comment]
It should only affect a single browser most likely as well. Without a repro we unfortunately can't act on it. We'd then bisect next the version and the code which is slower.

**Comment by MattEarth** [comment]
I'm aware of that... I'm trying to find the part which is causing the slowness... If I'll have clear case, I'll share the code.

**Comment by mxschmitt** [comment]
It could totally be, if your iframe loads multiple more iframes or does expensive operations in it, that might be related.

**Comment by MattEarth** [comment]
Hmm... still digging. I'm using a lot of waitForSelector method - I've feeling that something changed in the method...

**Comment by mxschmitt** [comment]
In theory you should never use this method, since all the web-first assertions which we have wait automatically: https://playwright.dev/docs/test-assertions

**Comment by MattEarth** [comment]
This is the theory... :D And it probably works in the simple webapps... Whereas my experience shows me, that's not enough in case of bigger app with iframes. I've the same code (running locally) with different pw versions... Observable difference is that, the newest version waits longer during gathering data for assertions. Already tried changing waitForSelectro to locator().isVisible()... Didn't help.

**Comment by mxschmitt** [comment]
If you have a use-case where it does not work, let us know! web-first assertions, frameLocators and locators are from our experience solving it.

So let us know if you have a project which you can share!

**Comment by MattEarth** [comment]
OK, so, I've found the root cause. It seems that between the versions, the following piece of code is running differentially: 
  this code in version 1.28 is working pretty well (without any issues):
`async getIframe(productName) {`
`  const iframe = await page.waitForSelector(selectors.productIFrame(productName), {`
    `    timeout: timeouts.TIMEOUT_60S,`
    `    state: 'visible',`
    `  });`
    `  const scope = await iframe.contentFrame();`
    `  return scope;`
`}`
Whereas in version 1.31 becomes very slow.... After changing it to the following: 
`async getIframe(productName) {`
`  const iframe = await page.$(selectors.productIFrame(productName));`
    `  const scope = await iframe.contentFrame();`
    `  return scope;`
`}`
It starts running smoothly on version 1.31, but it also starts failing on version 1.28 with the error 'TypeError: Cannot read property 'contentFrame' of null' (too fast execution on version 1.28).

**Comment by mxschmitt** [comment]
You are using ElementHandles which are racy by default, that explain the contentFrame of null error. You need to use [FrameLocators](https://playwright.dev/docs/api/class-framelocator). So general rule is to never call `$`, `$$`, `contentFrame`, etc. See here for [ElementHandles vs. Locators](https://playwright.dev/docs/handles#locator-vs-elementhandle). And make sure to use [web-first assertions](https://playwright.dev/docs/test-assertions) which wait by default, so you don't need waitForSelector. If you give me more code which uses the getIFrame I can help you rewriting it to FrameLocators.

**Comment by MattEarth** [comment]
For assertion, I'm using chai library... 
OK, I've changed the code to use frameLocator (seems to work in both versions). I also removed usage of waitForSelector method. Still two times slower. I really can't share the whole project, however, the simple case looks as following: 
1. log in to the 'portal'
2. gathering date using the method:
`static async getSupportContainerDetails(page) {`
`       const iframeContent = await elementsHelper.switchToIframeContext();`
`        const handle = iframeContent.locator(selectors.containerPage(page));`
`        const isDisplayed = await handle.isVisible();`
`        const headerText = await handle.locator(selectors.containerHeader).innerText();`
`        const icon = handle.locator(selectors.containerLinkImage);`
`        const iconLink = await icon.getAttribute('src');`
`        const doc = handle.locator(selectors.containerLink);`
`        const docLink = await doc.getAttribute('href');`
`        return { isDisplayed, headerText, iconLink, docLink };`
`    }`
3. The method is invoked for time on different subpages. 
4. couple expects and that's it. End of case. 
btw. After the changed the switchToIframeContext() method is like:
`const scope = page.frameLocator(selectors.productIFrame(productName));`
`return scope;`
Any advices/suggestions?

**Comment by mxschmitt** [comment]
Yeah Chai and Playwright's assertions are significant different, because our assertions retry until the condition is met, what Chai does not. (ours are specifically made for Playwright and end-to-end testing). This ends up in the state that chai or normal expect assertions are flaky and ours are not.

So it should just be:

- [`await expect(page.frameLocator(..).locator(..)).toBeVisible()`](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-visible)
- [`await expect(page.frameLocator(..).locator(..)).toHaveAttribute('src', '...')`](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-attribute)
- [`await expect(page.frameLocator(..).locator(..)).toHaveText('my-text')`](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-text)

Its unfortunately hard to tell from above what the issue is. I would recommend the following: make sure to use [best-practise](https://playwright.dev/docs/best-practices) assertions, migrate to [Playwright test-runner](https://playwright.dev/docs/intro) (you are missing out on [html report](https://playwright.dev/docs/trace-viewer-intro#opening-the-html-report), [horizontal/vertical scaling](https://playwright.dev/docs/test-parallel), better error reporting, [vscode extension support](https://playwright.dev/docs/getting-started-vscode) etc. and then see if you still experience issues with Playwright.

**Comment by MattEarth** [comment]
Hmm... Applied all your suggestions... Still 10s slower than in version 1.28 (without the changes - with all waitForSelectors and so on)... 
It's definitely something with handling the iframes or waitForSelector method. Nothing much to do here... I need to figure it out by my own, because I can't share the repo... Just referring to old code which is working pretty well with old version (1.28) -> Any suggestions what could change in the following methods between the versions:
* waitForSelector - maybe you added some additional checks?! 
* contentFrame

**Comment by dgozman** [comment]
@MattEarth Note that new Playwright version comes with new browsers. Perhaps you hit a rendering issue in the new browser? You can check it by using the same Google Chrome stable version with both Playwright versions - [specify `channel: 'chrome'`](https://playwright.dev/docs/browsers#google-chrome--microsoft-edge) in your config.

Other than that, I don't think we can help without being able to run your tests locally and profiling. Therefore, I am closing this issue. If you come up with a standalone repro that has this problem, please file a new issue with the repro, and we'll be happy to help!
